File: server/robot-server/src/game.rs
use std::{
    collections::{HashMap, HashSet},
    hash::Hash,
};

const SIZE: usize = 16;
const NUM_ROBOTS: usize = 4;

/// a wall is present between two tiles
type Connection = (Tile, Tile);

#[derive(Debug, Clone, Copy, Eq)]
pub struct Wall(Connection);

impl PartialEq for Wall {
    fn eq(&self, other: &Self) -> bool {
        let mut wall1: [Tile; 2] = self.0.into();
        let mut wall2: [Tile; 2] = other.0.into();
        wall1.sort();
        wall2.sort();
        wall1 == wall2
    }
}

impl Hash for Wall {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        let mut wall: [Tile; 2] = self.0.into();
        wall.sort();
        wall.hash(state);
    }
}

#[derive(Debug, Clone)]
pub struct Player {
    ip: String, // TODO: replace by websocket?
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct Tile {
    x: usize,
    y: usize,
}

impl Tile {
    fn new(x: usize, y: usize) -> Self {
        Self { x, y }
    }

    fn new_random() -> Self {
        Self {
            x: rand::random::<usize>() % SIZE,
            y: rand::random::<usize>() % SIZE,
        }
    }

    fn towards(&self, direction: Direction) -> Option<Tile> {
        match direction {
            Direction::Up => {
                let new_y = self.y.checked_sub(1)?;
                Some(Tile::new(self.x, new_y))
            }
            Direction::Down => {
                if self.y >= SIZE - 1 {
                    None
                } else {
                    Some(Tile::new(self.x, self.y + 1))
                }
            }
            Direction::Left => {
                let new_x = self.x.checked_sub(1)?;
                Some(Tile::new(new_x, self.y))
            }
            Direction::Right => {
                if self.x >= SIZE - 1 {
                    None
                } else {
                    Some(Tile::new(self.x + 1, self.y))
                }
            }
        }
    }

    fn wall_on(&self, direction: Direction) -> Wall {
        let other_tile = self.towards(direction).unwrap();
        Wall((self.clone(), other_tile))
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Direction {
    Up,
    Down,
    Left,
    Right,
}

use Direction::*;

impl Direction {
    fn all() -> [Self; 4] {
        [Up, Down, Left, Right]
    }
}

//
//
//

//   0 1 2 3 4 5 6 7 8 9 A B C D E F
// 0 . . . . .|. . . . . . .|. _ . .
// 1 . _ . . . . _|. . . . . .|. . .
// 2 .|. . . . . . . . _|. . . . . .
// 3 . . . . . . . . . . . . . . . _
// 4 . . . . . . _ . . . . . . . . .
// 5 _ . . . . . .|. . . . _ . .|_ .
// 6 . . .|_ . . . . . . . .|. . . .
// 7 . . . . . . . x x . . . _ . . .
// 8 . . _ . .|_ . x x . . .|. . . .
// 9 . .|. . . . . . . .|_ . . . . .
// A . . . . . . . . . . . . . . . _
// B _ . . . . . . . . _|. . . . _ .
// C . . . . _ . . . . . . . . . .|.
// D . . . . .|. . . . . . . . _ . .
// E . _|. . . . . . . . . . .|. . .
// F . . . . . .|. . . . .|. . . . .
const CORNER_WALLS: [(Tile, (Direction, Direction)); 17] = [
    (Tile { x: 1, y: 2 }, (Up, Left)),
    (Tile { x: 12, y: 8 }, (Up, Left)),
    (Tile { x: 2, y: 9 }, (Up, Left)),
    (Tile { x: 13, y: 14 }, (Up, Left)),
    (Tile { x: 13, y: 1 }, (Up, Left)),
    (Tile { x: 6, y: 5 }, (Up, Right)),
    (Tile { x: 4, y: 13 }, (Up, Right)),
    (Tile { x: 14, y: 12 }, (Up, Right)),
    (Tile { x: 11, y: 6 }, (Up, Right)),
    (Tile { x: 6, y: 1 }, (Down, Right)),
    (Tile { x: 9, y: 2 }, (Down, Right)),
    (Tile { x: 9, y: 11 }, (Down, Right)),
    (Tile { x: 1, y: 14 }, (Down, Right)),
    (Tile { x: 3, y: 6 }, (Down, Left)),
    (Tile { x: 14, y: 5 }, (Down, Left)),
    (Tile { x: 10, y: 9 }, (Down, Left)),
    (Tile { x: 5, y: 8 }, (Down, Left)),
];

const NORMAL_WALLS: [(Tile, Direction); 8] = [
    //top
    (Tile { x: 4, y: 0 }, Right),
    (Tile { x: SIZE - 4, y: 0 }, Left),
    // bottom
    (Tile { x: 6, y: 15 }, Right),
    (Tile { x: SIZE - 5, y: 15 }, Left),
    // left
    (Tile { x: 0, y: 6 }, Down),
    (Tile { x: 0, y: SIZE - 4 }, Up),
    // right
    (Tile { x: 15, y: 4 }, Down),
    (Tile { x: 15, y: SIZE - 5 }, Up),
];

//
//
//

pub struct Board {
    robots: Vec<Tile>,
    robots_backup_positions: Vec<Tile>,
    walls: HashMap<Wall, bool>,
    target: Tile,
    targetForRobot: usize,
}

impl Board {
    fn new() -> Self {
        Self {
            robots: Vec::new(),
            robots_backup_positions: Vec::new(),
            walls: HashMap::new(),
            target: Tile { x: 0, y: 0 },
            targetForRobot: 0,
        }
    }

    fn add_wall(&mut self, wall: Wall) {
        self.walls.insert(wall, true);
    }

    fn init_default(&mut self) {
        // place the walls to form the box in the middle
        self.add_wall(Tile::new(8, 7).wall_on(Right));
        self.add_wall(Tile::new(6, 8).wall_on(Right));
        self.add_wall(Tile::new(8, 8).wall_on(Right));
        self.add_wall(Tile::new(7, 6).wall_on(Down));
        self.add_wall(Tile::new(7, 8).wall_on(Down));
        self.add_wall(Tile::new(8, 6).wall_on(Down));
        self.add_wall(Tile::new(8, 8).wall_on(Down));

        // place 8 normal walls
        for (tile, direction) in NORMAL_WALLS.iter() {
            self.add_wall(tile.wall_on(*direction));
        }

        // place the corner walls
        for (tile, (dir1, dir2)) in CORNER_WALLS.iter() {
            self.add_wall(tile.wall_on(*dir1));
            self.add_wall(tile.wall_on(*dir2));
        }
    }

    fn init_robots(&mut self) {
        self.robots.clear();

        let mut occupied = HashSet::new();
        loop {
            if occupied.len() == NUM_ROBOTS {
                break;
            }

            let tile = Tile::new_random();
            if occupied.contains(&tile) || self.in_middle_box(tile) {
                continue;
            }
            occupied.insert(tile);
            self.robots.push(tile);
        }

        self.robots_backup_positions = self.robots.clone();
    }

    fn clear(&mut self) {
        self.robots.clear();
        self.robots_backup_positions.clear();
    }

    fn reset(&mut self) {
        self.robots = self.robots_backup_positions.clone();
    }

    fn has_wall(&self, curr_tile: Tile, direction: Direction) -> bool {
        let other_tile = if let Some(tile) = curr_tile.towards(direction) {
            tile
        } else {
            return true;
        };
        let key = Wall((curr_tile, other_tile));
        self.walls.get(&key).copied().unwrap_or(false)
    }

    fn has_robot(&self, tile: Tile) -> Option<usize> {
        self.robots.iter().position(|t| t == &tile)
    }

    fn in_middle_box(&self, tile: Tile) -> bool {
        tile.x > 6 && tile.x < 9 && tile.y > 6 && tile.y < 9
    }

    fn can_move_robot(&self, robot: usize) -> Vec<Tile> {
        let curr_tile = self.robots[robot];
        let mut possible_tiles = vec![];

        for direction in Direction::all() {
            let mut final_tile = curr_tile;
            loop {
                let next_tile = if let Some(tile) = final_tile.towards(direction) {
                    tile
                } else {
                    break;
                };
                if self.has_wall(final_tile, direction)
                    || self.has_robot(next_tile).is_some()
                    || self.in_middle_box(next_tile)
                {
                    break;
                }
                final_tile = next_tile;
            }
            if final_tile != curr_tile {
                possible_tiles.push(final_tile);
            }
        }
        possible_tiles
    }

    fn move_robot(&mut self, robot: usize, tile: Tile) -> Result<bool, &'static str> {
        // check if movement is valid
        if !self.can_move_robot(robot).contains(&tile) {
            return Err("invalid move");
        }

        self.robots[robot] = tile;

        Ok(false)
    }
}

pub struct Game {
    board: Board,
    players: Vec<Player>,
}

impl Game {
    pub fn new() -> Self {
        let mut board = Board::new();
        board.init_default();

        Self {
            board,
            players: Vec::new(),
        }
    }

    pub fn add_player(&mut self, player: Player) {
        self.players.push(player);
    }

    pub fn start_game(&mut self) {
        self.board.clear();
        self.board.init_robots();
        self.start_next_round();
    }

    pub fn start_next_round(&mut self) {
        loop {
            /* const pos = Math.floor(Math.random() * corner_walls.length);
            const tile = corner_walls[pos][0];
            if (this.board.hasRobot(tile) === null) {
              this.board.target = tile;
              break;
            } */
            let pos = rand::random::<usize>() % CORNER_WALLS.len();
            let tile = CORNER_WALLS[pos].0;
            if self.board.has_robot(tile).is_none() {
                self.board.target = tile;
                break;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_walls() {
        let wall1 = Wall((Tile::new(1, 2), Tile::new(2, 2)));
        let wall2 = Wall((Tile::new(2, 2), Tile::new(1, 2)));
        assert_eq!(wall1, wall2);
        let mut hashmap = HashMap::new();
        hashmap.insert(wall1, true);
        assert_eq!(hashmap.get(&wall2), Some(&true));
    }
}

File: shared/game.ts
// Ricochet Robots core game logic — pure, no DOM, shared by server (authoritative)
// and client (private attempt board + move preview).

export const SIZE = 16;
export const NUM_ROBOTS = 4; // must match ROBOT_COLORS length

export type Tile = { x: number; y: number };

export type Direction = "up" | "down" | "left" | "right";
export const DIRECTIONS: Direction[] = ["up", "down", "left", "right"];

export const ROBOT_COLORS = ["red", "blue", "green", "purple"] as const;
export type RobotColor = (typeof ROBOT_COLORS)[number];

// A move slides a given robot in a direction until it stops.
export type Move = { robot: number; dir: Direction };

// The minimal puzzle state that travels over the wire. Walls are deterministic
// (init_default) so they are reconstructed on both ends and never serialized.
export type Puzzle = {
  robotStarts: Tile[];
  target: Tile;
  targetForRobot: number;
};

// a wall sits between two adjacent tiles
type Connection = [Tile, Tile];

//   0 1 2 3 4 5 6 7 8 9 A B C D E F
// 0 . . . . .|. . . . . . .|. _ . .
// 1 . _ . . . . _|. . . . . .|. . .
// 2 .|. . . . . . . . _|. . . . . .
// 3 . . . . . . . . . . . . . . . _
// 4 . . . . . . _ . . . . . . . . .
// 5 _ . . . . . .|. . . . _ . .|_ .
// 6 . . .|_ . . . . . . . .|. . . .
// 7 . . . . . . . x x . . . _ . . .
// 8 . . _ . .|_ . x x . . .|. . . .
// 9 . .|. . . . . . . .|_ . . . . .
// A . . . . . . . . . . . . . . . _
// B _ . . . . . . . . _|. . . . _ .
// C . . . . _ . . . . . . . . . .|.
// D . . . . .|. . . . . . . . _ . .
// E . _|. . . . . . . . . . .|. . .
// F . . . . . .|. . . . .|. . . . .
const CORNER_WALLS: [Tile, string][] = [
  [{ x: 1, y: 2 }, "top-left"],
  [{ x: 12, y: 8 }, "top-left"],
  [{ x: 2, y: 9 }, "top-left"],
  [{ x: 13, y: 14 }, "top-left"],
  [{ x: 13, y: 1 }, "top-left"],
  [{ x: 6, y: 5 }, "top-right"],
  [{ x: 4, y: 13 }, "top-right"],
  [{ x: 14, y: 12 }, "top-right"],
  [{ x: 11, y: 6 }, "top-right"],
  [{ x: 6, y: 1 }, "bottom-right"],
  [{ x: 9, y: 2 }, "bottom-right"],
  [{ x: 9, y: 11 }, "bottom-right"],
  [{ x: 1, y: 14 }, "bottom-right"],
  [{ x: 3, y: 6 }, "bottom-left"],
  [{ x: 14, y: 5 }, "bottom-left"],
  [{ x: 10, y: 9 }, "bottom-left"],
  [{ x: 5, y: 8 }, "bottom-left"],
];

const NORMAL_WALLS: [Tile, Direction][] = [
  // top
  [{ x: 4, y: 0 }, "right"],
  [{ x: SIZE - 4, y: 0 }, "left"],
  // bottom
  [{ x: 6, y: 15 }, "right"],
  [{ x: SIZE - 5, y: 15 }, "left"],
  // left
  [{ x: 0, y: 6 }, "down"],
  [{ x: 0, y: SIZE - 4 }, "up"],
  // right
  [{ x: 15, y: 4 }, "down"],
  [{ x: 15, y: SIZE - 5 }, "up"],
];

export function randInt(max: number): number {
  return Math.floor(Math.random() * max);
}

export class Board {
  robots: Tile[] = [];
  robotsBackupPositions: Tile[] = [];
  walls: Map<string, boolean> = new Map();
  target: Tile = { x: 0, y: 0 };
  targetForRobot = 0;

  reset() {
    this.robots = this.robotsBackupPositions.map((t) => ({ ...t }));
  }

  otherTileFrom(tile: Tile, direction: Direction): Tile {
    const res = { ...tile };
    if (direction === "up") {
      if (tile.y <= 0) throw new Error("edge");
      res.y -= 1;
    } else if (direction === "down") {
      if (tile.y >= SIZE - 1) throw new Error("edge");
      res.y += 1;
    } else if (direction === "left") {
      if (tile.x <= 0) throw new Error("edge");
      res.x -= 1;
    } else if (direction === "right") {
      if (tile.x >= SIZE - 1) throw new Error("edge");
      res.x += 1;
    }
    return res;
  }

  // Returns the tile a robot lands on if it slides in `direction`, or null if it
  // cannot move at all in that direction.
  slide(robot: number, direction: Direction): Tile | null {
    const curr = this.robots[robot];
    let finalTile = curr;
    while (true) {
      let next: Tile;
      try {
        next = this.otherTileFrom(finalTile, direction);
      } catch {
        break;
      }
      if (
        this.hasWall(finalTile, direction) ||
        this.hasRobot(next) !== null ||
        this.inMiddleBox(next)
      ) {
        break;
      }
      finalTile = next;
    }
    if (finalTile.x === curr.x && finalTile.y === curr.y) return null;
    return finalTile;
  }

  // All tiles the robot could slide to (one per direction it can move).
  canMoveRobot(robot: number): Tile[] {
    const tiles: Tile[] = [];
    for (const direction of DIRECTIONS) {
      const dest = this.slide(robot, direction);
      if (dest) tiles.push(dest);
    }
    return tiles;
  }

  // Moves a robot to a destination tile. Returns true if the win condition is met.
  moveRobot(robot: number, tile: Tile): boolean {
    this.robots[robot] = tile;
    return (
      tile.x === this.target.x &&
      tile.y === this.target.y &&
      robot === this.targetForRobot
    );
  }

  hasRobot(tile: Tile): number | null {
    for (let idx = 0; idx < this.robots.length; idx++) {
      if (this.robots[idx].x === tile.x && this.robots[idx].y === tile.y) {
        return idx;
      }
    }
    return null;
  }

  private wallKey(connection: Connection): string {
    connection.sort((a, b) => (a.x - b.x !== 0 ? a.x - b.x : a.y - b.y));
    return `(${connection[0].x},${connection[0].y})-(${connection[1].x},${connection[1].y})`;
  }

  hasWall(curr: Tile, direction: Direction): boolean {
    try {
      const other = this.otherTileFrom(curr, direction);
      return this.walls.get(this.wallKey([curr, other])) || false;
    } catch {
      return true; // board edge counts as a wall
    }
  }

  addWall(curr: Tile, direction: Direction) {
    const other = this.otherTileFrom(curr, direction);
    this.walls.set(this.wallKey([curr, other]), true);
  }

  addCornerWall(curr: Tile, corner: string) {
    if (corner === "top-left") {
      this.addWall(curr, "up");
      this.addWall(curr, "left");
    } else if (corner === "top-right") {
      this.addWall(curr, "up");
      this.addWall(curr, "right");
    } else if (corner === "bottom-left") {
      this.addWall(curr, "down");
      this.addWall(curr, "left");
    } else if (corner === "bottom-right") {
      this.addWall(curr, "down");
      this.addWall(curr, "right");
    } else {
      throw new Error("invalid corner");
    }
  }

  init_default() {
    // box in the middle
    this.addWall({ x: 8, y: 7 }, "right");
    this.addWall({ x: 6, y: 8 }, "right");
    this.addWall({ x: 8, y: 8 }, "right");
    this.addWall({ x: 7, y: 6 }, "down");
    this.addWall({ x: 7, y: 8 }, "down");
    this.addWall({ x: 8, y: 6 }, "down");
    this.addWall({ x: 8, y: 8 }, "down");

    for (const [tile, direction] of NORMAL_WALLS) {
      this.addWall(tile, direction);
    }
    for (const [tile, corner] of CORNER_WALLS) {
      this.addCornerWall(tile, corner);
    }
  }

  inMiddleBox(tile: Tile): boolean {
    return (
      (tile.x === 7 && (tile.y === 7 || tile.y === 8)) ||
      (tile.x === 8 && (tile.y === 7 || tile.y === 8))
    );
  }

  initRobots() {
    this.robots = [];
    const occupied = new Set<string>();
    while (occupied.size < NUM_ROBOTS) {
      const x = randInt(SIZE);
      const y = randInt(SIZE);
      const key = `${x},${y}`;
      const tile = { x, y };
      if (occupied.has(key) || this.inMiddleBox(tile)) continue;
      occupied.add(key);
      this.robots.push(tile);
    }
    this.robotsBackupPositions = this.robots.map((t) => ({ ...t }));
  }

  // Load a fixed puzzle (used by the client and validator).
  loadPuzzle(puzzle: Puzzle) {
    this.robots = puzzle.robotStarts.map((t) => ({ ...t }));
    this.robotsBackupPositions = puzzle.robotStarts.map((t) => ({ ...t }));
    this.target = { ...puzzle.target };
    this.targetForRobot = puzzle.targetForRobot;
  }
}

export class Game {
  board: Board;

  constructor() {
    this.board = new Board();
    this.board.init_default();
  }

  // Place robots and pick the first target.
  startGame() {
    this.board.initRobots();
    this.startNextRound();
  }

  // Pick a fresh random target (a corner-wall tile not under a robot) and a
  // random robot to be sent there, then snapshot robot positions as the puzzle.
  startNextRound() {
    while (true) {
      const tile = CORNER_WALLS[randInt(CORNER_WALLS.length)][0];
      if (this.board.hasRobot(tile) === null) {
        this.board.target = { ...tile };
        break;
      }
    }
    this.board.targetForRobot = randInt(NUM_ROBOTS);
    this.board.robotsBackupPositions = this.board.robots.map((t) => ({ ...t }));
  }

  puzzle(): Puzzle {
    return {
      robotStarts: this.board.robots.map((t) => ({ ...t })),
      target: { ...this.board.target },
      targetForRobot: this.board.targetForRobot,
    };
  }
}

// Build a board with the default walls and a loaded puzzle. Handy for clients.
export function boardFromPuzzle(puzzle: Puzzle): Board {
  const board = new Board();
  board.init_default();
  board.loadPuzzle(puzzle);
  return board;
}

export type SolutionResult = {
  valid: boolean; // every move was a legal slide
  reachedTarget: boolean; // ended on the target with the right robot
  moveCount: number; // number of slides that actually happened
};

// Authoritatively replay an ordered list of moves from the puzzle start and
// report whether it legally reaches the target, and in how many moves. Used by
// the server so clients can never spoof a move count.
export function validateSolution(puzzle: Puzzle, moves: Move[]): SolutionResult {
  const board = boardFromPuzzle(puzzle);
  let moveCount = 0;
  for (const move of moves) {
    if (
      move == null ||
      move.robot < 0 ||
      move.robot >= board.robots.length ||
      !DIRECTIONS.includes(move.dir)
    ) {
      return { valid: false, reachedTarget: false, moveCount };
    }
    const dest = board.slide(move.robot, move.dir);
    if (!dest) {
      // robot can't move that way — illegal move
      return { valid: false, reachedTarget: false, moveCount };
    }
    moveCount++;
    const won = board.moveRobot(move.robot, dest);
    if (won) {
      return { valid: true, reachedTarget: true, moveCount };
    }
  }
  return { valid: true, reachedTarget: false, moveCount };
}
